site_scons/site_tools/antlr4.py

```diff
@@ -0,0 +1,102 @@
+import os
+import SCons.Util
+from SCons.Script import Builder, Action, Dir, File, Entry
+
+
+def _detect(env):
+    """Try to find the Antlr parser generator"""
+    try:
+        return env['ANTLR4']
+    except KeyError:
+        raise SCons.Errors.StopError(
+            'Could not find antlr4 parser generator')
+
+
+def _antlr4_emitter(target, source, env):
+    """Process sources and flags"""
+    target = []
+    antlr4_suffix = env.subst('$ANTLR4_SUFFIX')
+    antlr4_h_suffix = env.subst('$ANTLR4_HSUFFIX')
+    antlr4_cc_suffix = env.subst('$ANTLR4_CCSUFFIX')
+    antlr4_token_suffix = env.subst('$ANTLR4_TOKENSUFFIX')
+    if env['ANTLR4_FLAGS']:
+        antlr4flags = env.subst('$ANTLR4_FLAGS', target=target, source=source)
+        flags = SCons.Util.CLVar(antlr4flags)
+    else:
+        flags = SCons.Util.CLVar('')
+
+    # -o flag
+    if env['ANTLR4_OUT']:
+        env['ANTLR4_OUT'] = Dir(env['ANTLR4_OUT'])
+        flags.append('-o ${ANTLR4_OUT}')
+
+    # -lib flag
+    if env['ANTLR4_GLIB']:
+        env['ANTLR4_GLIB'] = Dir(env['ANTLR4_GLIB'])
+        flags.append('-lib ${ANTLR4_GLIB}')
+
+    # update antlr4 flags
+    env['ANTLR4_FLAGS'] = str(flags)
+
+    # compute targets
+    deps = []
+    grammars = [] 
+
+    lexerFiles = (('', (antlr4_h_suffix, antlr4_cc_suffix, antlr4_token_suffix)),)
+    parserFiles = (('', (antlr4_h_suffix, antlr4_cc_suffix)),
+                   ('BaseListener', (antlr4_h_suffix, antlr4_cc_suffix)),
+                   ('Listener', (antlr4_h_suffix, antlr4_cc_suffix)))
+
+    for src in source:
+        src = File(src)
+        stem = src.abspath
+        if stem.endswith(antlr4_token_suffix):
+            deps.append(src)
+            continue
+        grammars.append(src)
+        if stem.endswith(antlr4_suffix):
+            stem = stem[:-len(antlr4_suffix)]
+        kindInfo = lexerFiles if "Lexer" in stem else parserFiles
```

Comment:
Add comment here stating explicitly that it relies on a particular naming convention of g4 files?